### PR TITLE
ADC: Rewrite to use template class

### PR DESCRIPTION
**Pull request by Hagridpon3**

### CODE REVIEW

File: src/common/adc.h [hunk 1]
```diff
@@ -3,17 +3,57 @@
...
-extern uint16_t get_adc_channel_value(ADC_HandleTypeDef *adc, uint32_t channel);
+template <ADC_HandleTypeDef &adc, size_t channels>
+class AdcDma {
+public:
+    AdcDma()
+        : m_data() {};
+    void init() {
+#ifndef SIM_HEATER
+        HAL_ADC_Init(&adc);
+        HAL_ADC_Start_DMA(&adc, (uint32_t *)m_data, channels); // Start ADC in DMA mode and
```
Comment (mkbel) [inline src/common/adc.h:18]: Suggest: Use c++ style cast (`static_cast`) instead of c-style cast.

**Comment by mkbel** [comment]
Maybe it would be worth to rename adc.h to adc.hpp, as it can not be used in C.

File: src/common/adc.h (hunk 1 above)
Comment (Hagridpon3) [reply src/common/adc.h]: Solved

**Review by mkbel — DISMISSED** [review]

**Comment by mkbel** [comment]
What positive effect is gained by added complexity of interface class?

**Review by RadekVana — APPROVED** [review]

**Review by mkbel — APPROVED** [review]

**Comment by RadekVana** [comment]
I don't like `SIM_HEATER` macro and would prefer not having it in header files and in future replaced by just different sources in cmake.

**Pull request merged**